### PR TITLE
Proposal: pull request

### PR DESCRIPTION
**Pull request by nopcoder**
Initial mechanism to have manuall approved pull-request - introduce changes into a branch.

### CODE REVIEW

**Review by N-o-Z — COMMENTED** [review]
Very exciting!
Do we need explicit authorization for PR approval?
Should we allow merge (without force) when no reviewers are assigned?
PR update == commit / revert / rebase / etc. It means we have to add a hook trigger on all flows? Will the PR merge trigger a secondary hook call (pre/post merge) in addition to the PR hook?

File: design/open/pull-request.md
Comment (itaiad200) [inline design/open/pull-request.md:19]:
```suggestion
PR will not be updated with changes to the source or destination branches, like updates to objects, commits, merges, etc. 
```

File: design/open/pull-request.md
Comment (itaiad200) [inline design/open/pull-request.md:20]: That's tricky. I might be reviewing a PR that has no chances to be merged since there are conflicts.

File: design/open/pull-request.md
Comment (itaiad200) [inline design/open/pull-request.md:22]: How would branch protection rules be updated to support pull requests?

File: design/open/pull-request.md
Comment (itaiad200) [inline design/open/pull-request.md:9]: I don't think `"discuss"` is something that is supported by the below proposal.

Comment (itaiad200) [inline design/open/pull-request.md:9]: Just merges for now, right?
```suggestion
- Mechanism to review, discuss and approve merges before applying to a branch
```

File: design/open/pull-request.md
Comment (itaiad200) [inline design/open/pull-request.md:10]: I think we need to elaborate on it. `pull_request_open` trigger with actions that must pass?

File: design/open/pull-request.md
Comment (nopcoder) [reply design/open/pull-request.md:20]: We can enable client side request to run dry merge - but currelty we don't know if a merge operation will have conflicts until we do the actual merge.

I think it rases another question - does a PR represent the merge itself or the request?
If we have a pending request to merge a commit or a branch, any future request to merge this PR should point to the same PR? Can we have multiple PRs requesting the same thing open in parallel? (will move this one to issues to discuss)

File: design/open/pull-request.md
Comment (nopcoder) [reply design/open/pull-request.md:22]: Current branch protection enable merge only changes - as PR performs just merge we can use it to push changes into a branch just by merge/PRs.

Additional support can be added to the branch protection by allow only through PRs. And more options like - only approved, with specific checks and etc.

File: design/open/pull-request.md
Comment (nopcoder) [reply design/open/pull-request.md:9]: Added discussion section that will include zero or more comments. Each comment will be added to the end of the discussion.

**Comment by nopcoder** [comment]
Very exciting!
> I can't wait I think it is the next step for enabling better workflows using lakeFSt

Do we need explicit authorization for PR approval?
> Same as in merge - currently we check create commit on target branch

Should we allow merge (without force) when no reviewers are assigned?
> Can open to discussion - I think we can add future 'commit as admin' option. Or enforce at leave X approval for merge PRs on branch part of branch protection.

PR update == commit / revert / rebase / etc. It means we have to add a hook trigger on all flows? Will the PR merge trigger a secondary hook call (pre/post merge) in addition to the PR hook?
> The current proposal does not include pull-request update as part of branch update. We can include this functionality (open to discussion). But I don't think it should effect the actions we currently triggering and the pull-request event that we will trigger as part of this feature.
The user will be able to configure what will run as part of the pull-request step and if still there is a pre-merge hook - it can still block the pull-request 'Merge' operation.

File: design/open/pull-request.md
Comment (nopcoder) [reply design/open/pull-request.md:9]: Not going to change the support with pre-merge - it will trigger when you merge the PR as it is equal to a merge operation.
The PR trigger 'pull-request' action after internal merge operation is done without conflicts.
The new event event will enable check the changes before approving the merge.

File: design/open/pull-request.md
Comment (nopcoder) [reply design/open/pull-request.md:10]: Added some description on the action - currently I think that an action 'pull-request' that will be triggered on creation and by request is enough. Until we support updates.

File: design/open/pull-request.md
Comment (itaiad200) [inline design/open/pull-request.md:25]: What does `by request` mean? Can I rerun an action of a PR? Can I request a trigger for PR that did not run in the PR?

File: design/open/pull-request.md
Comment (itaiad200) [inline design/open/pull-request.md:33]: This can be done automatically, simply diff the 2 and notify on conflict. btw I think that the UI will present the diff and will show conflicts anyhow

File: design/open/pull-request.md
Comment (nopcoder) [reply design/open/pull-request.md:25]: A request to re-run the action as the target branch can hold/update over time. We will need to enable re-running the action as it verify the merge with the target at a point in time.

**Review by itaiad200 — APPROVED** [review]

File: design/open/pull-request.md
Comment (nopcoder) [reply design/open/pull-request.md:33]: Make sense I'll remove this item, thanks.

File: design/open/pull-request.md
Comment (N-o-Z) [inline design/open/pull-request.md:14]: What is the expected behavior if src or dst were deleted after the PR was opened?

**Review by N-o-Z — APPROVED** [review]
Thanks, when do we start? 😄

File: design/open/pull-request.md
Comment (nopcoder) [reply design/open/pull-request.md:14]: Good point - I'll update the proposal to act only on active branches. When you delete a branch the associated pull-requests should be closed.

Comment (N-o-Z) [reply design/open/pull-request.md:14]: If a Branch is deleted, do we need to go over the list of PR and check if it is associated to any? Or do we close the PR only on a failed Merge attempt?
If we want to couple it, maybe we should put the PRs under the dst branch? This way If a src branch is deleted, the PR will be deleted as part of it.
Also, I think we can ignore deletion of src branch and handle the PR as an independent from it.

**Comment by nopcoder** [comment]
Resolve open items. Main change includes:
1. PR is per source/target branch
2. PR is not for a specific commit
3. PR actions will be triggered on PR creation and on every commit to the source branch.

File: design/open/pull-request.md
Comment (nopcoder) [reply design/open/pull-request.md:14]: We will require to close all PRs related to the branch - as a source or destination - the idea is to represent a valid request.

Failed merge will keep the PR open and notify the requester that the PR failed to merge.

**Pull request merged**
